server/src/main/java/org/elasticsearch/cluster/metadata/TemplateUpgradeService.java

```diff
@@ -74,7 +71,8 @@
 
     public final Client client;
 
-    private final AtomicInteger updatesInProgress = new AtomicInteger();
+    final AtomicInteger upgradesInProgress = new AtomicInteger();
+    private volatile boolean allUpgradesSuccessful;
```

Comment:
Keep track if all template changes have been acknowledged. Retrying the template upgrade is expected if not all changes are successful.

Reply:
I'm not a big fan of this field.
It feels like it could just get pushed down to be an `AtomicBoolean` inside `upgradeTemplates`, or if we need it to be a field, I think it works better if its meaning is reversed e.g. `detectedUpgradeErrors`
As it stands we reset it to `true` even if we know something failed, which just feels wrong.

Reply:
++

Reply:
agreed, I have pushed it down as a local variable and made it flag failures: "It's the little things..."